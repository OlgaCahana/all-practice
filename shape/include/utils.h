#ifndef __UTILS_H__
#define __UTILS_H__

#include <stdio.h>


/*******************************************************************************/
/* CPP */
#if __cplusplus > 201102L
    #define NOEXCEPT noexcept
#else
    #define NOEXCEPT throw()
#endif

/******************************************************************************/
/* MIN2 MAX2 ABS */
#define MIN2(x, y) ((x < y) ? x : y)
#define MAX2(x, y) ((x > y) ? x : y)
#define ABS(x) ((x) < 0 ? ((x) * -1) : (x))

/**********************Failure handle*****************************************/
#define LOG(MSG) fprintf(stderr, "msg: %s \n\tfile: %s line: %d compiled at: %s\n",	\
	                 			 #MSG,      __FILE__, __LINE__,      __TIME__); 

#define UNUSED(param) ((void)param)

/* DEADMEM is the go-to bad address: */
#define DEAD_MEM(TYPE) ((TYPE)0xBADC0FFEE0DDF00D)

/* general "return if bad macro: " */
#define RETURN_IF_BAD(IS_GOOD, RET, MSG)                             \
if(!IS_GOOD)                                                         \
{                                                                    \
	fprintf(stderr, MSG);                                            \
	fprintf(stderr, " expression: %s has failed\n", #IS_GOOD);       \
	fprintf(stderr, "file: %s line: %d compiled at: %s\n",           \
	                    __FILE__, __LINE__,      __TIME__);          \
	return RET;                                                      \
}  


#define CHECK_ALLOC_IFBAD_FREE_RETURN(x, FREE, FREE2, RETURN_VAL)	\
do 											 	\
{										 		\
	if (NULL == x)						 		\
	{											\
		fprintf(stderr, "Allocation failed\n");	\
		free(FREE);						 		\
		free(FREE2);						 	\
		return RETURN_VAL;				 		\
	}									 		\
}										 		\
while(0)


#define CHECK_ALLOC(x, FREE, FREE2, RETURN_VAL)	\
do 											 	\
{										 		\
	if (NULL == x)						 		\
	{											\
		fprintf(stderr, "Allocation failed\n");	\
		free(FREE);						 		\
		free(FREE2);						 	\
		return RETURN_VAL;				 		\
	}									 		\
}										 		\
while(0)



/* for backwards compatability only, do not use in new files */
#define CHECK_IF_BAD(x, BAD_RESULT, FREE, FREE2, RETURN_VAL)	\
do 											 	\
{										 		\
	if (x == BAD_RESULT)						 		\
	{											\
		fprintf(stderr, "Got Bad Result\n");	\
		free(FREE);						 		\
		free(FREE2);						 	\
		return RETURN_VAL;				 		\
	}									 		\
}										 		\
while(0)

/*******************************DEBUG**********************************************/
#ifndef NDEBUG
	#define DEBUG_ONLY(exp) exp;
	#define DEBUG_ONLY_ARG(exp) ,exp
#else
	#define DEBUG_ONLY(exp) 
	#define DEBUG_ONLY_ARG(exp)
#endif /* NDEBUG */

#if __cplusplus<201104
#define STATIC_ASSERT(EXP, NAME) enum {NAME = 1/EXP};
#else
#define STATIC_ASSERT(EXP, NAME) static_assert(EXP);
#endif /* cplusplus<201104 */


#endif /*__UTILS_H__*/
