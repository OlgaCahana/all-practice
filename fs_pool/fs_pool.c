/*******************************************************************************
*	Written by:		Olga Danielle Cahana									   *
*	Date:			17/01/2022												   *
*	Reviewed by:	Meirav						    					 	   *
*																			   *
*******************************************************************************/
#include <stddef.h> /* size_t */
#include <assert.h>

#include "fs_pool.h"    /* fs_pool types and function declerations */
#include "utils.h"       /* util macros */


struct fixed_size_pool
{
    size_t first_free_blk;     /* offset in bytes from adress of stuct to available chunk */
    DEBUG_ONLY(size_t pool_size)   /* dbg? keeps number of elements - for free function (dbg) */
};


/****************************** Helpers ****************************************/
/* calculate the offset of the first chunk in the memory pool */
size_t CalcPoolOffset(size_t chunk_size);


/*************************API Implementaion*************************************/

fsp_ty *FSPoolInit(void *memory, size_t memory_size, size_t chunk_size)
{   
    fsp_ty *fsp_mem = memory;
    char *mem_curr = memory;
    size_t offset = CalcPoolOffset(chunk_size);

    /* asserts */
    assert(memory);
    assert(memory_size > chunk_size);
    assert(chunk_size >= sizeof(size_t));
    
    /* if needed to normalize chunk_size to be word aligned */
    /*chunk_size += sizeof(size_t) - chunk_size % sizeof(size_t);*/

    /* initialize metadata:    */
    fsp_mem->first_free_blk = offset;
    DEBUG_ONLY(fsp_mem->pool_size = memory_size);


    /* ptomote the mem_curr to the first non metadata block */
    mem_curr += offset;

    /*  eager initialization */
    /*  initialize each freelist's index content with the offset of the next free,
        the final offset is 0. */
    
    while(mem_curr + chunk_size < (char *)memory + memory_size)
    {
        offset += chunk_size;
        *((size_t *)mem_curr) = offset;
        mem_curr += chunk_size;
    }

    *((size_t *)mem_curr) = 0;
    return fsp_mem;
}


void *FSPoolAlloc(fsp_ty *fs_pool)
{
    size_t offset = 0;

    /*  handle empty freelist */
    if(fs_pool->first_free_blk == 0)
    {
        printf("No memory left in the pool!\n");
        return NULL;
    }
    
    /*  save freelist[first_blk] address in local temp var */
    offset = fs_pool->first_free_blk;

    /*  copy first_blk */

    /*  progression: first_blk = next */
    fs_pool->first_free_blk = *((char *)fs_pool + fs_pool->first_free_blk);

    /*  convert copy to pointer and return */
    return (char *)fs_pool + offset;
}


void FSPoolFree(fsp_ty *fs_pool, void *chunk_to_free)
{
    size_t offset = 0;
    /*  assert given pointer is part of the pool"
        chunk >= freelist && chunk <= (char *)freelist + pool_size  */
    assert(fs_pool);
    assert(chunk_to_free); 
    assert((char *)chunk_to_free > (char *)fs_pool);
    assert((char *)chunk_to_free < ((char *)fs_pool + fs_pool->pool_size));

    /* set next free offset of chunk_to free to current next free */
    *((size_t *)chunk_to_free) = fs_pool->first_free_blk;

    /* calculate the offset of chunk_to_free */
    offset = (char *)chunk_to_free - (char *)fs_pool;

    /* point first_blk at chunk */
    fs_pool->first_free_blk = offset;
}


size_t FSPoolCountFreeChunks(fsp_ty *fs_pool)
{ 
    /*  create counter. */
    size_t counter = 0;
    size_t curr_offset = 0;

    assert(fs_pool);

    /* save first_blk to temp variable */
    curr_offset = fs_pool->first_free_blk;

    /*  while temp isnt containing distinct value(0) - go to that address */
    while(curr_offset != 0)
    {
        ++counter;
        curr_offset = *(size_t *)((char *)fs_pool + curr_offset);
    }
       
    return counter;
}


size_t FSPoolCalcSize(size_t num_chunks, size_t chunk_size)
{
    assert(chunk_size > sizeof(size_t));
    
    /*  calculate size of num_chunks * chunk_size + size of fsp_ty alligned to chunk_size */
    return CalcPoolOffset(chunk_size) + num_chunks * chunk_size;
}



/****************************Helpers Impl*****************************************/
size_t CalcPoolOffset(size_t chunk_size)
{
    if(sizeof(fsp_ty) <= chunk_size)
    {
        return chunk_size;
    }
    
    return (sizeof(fsp_ty) / chunk_size)*chunk_size + 
            chunk_size * !!(sizeof(fsp_ty) % chunk_size);

}

