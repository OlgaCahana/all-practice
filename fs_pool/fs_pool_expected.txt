Test release FSPoolCalcSize. 		Test Passed
Test full FSPoolCountFreeChunks. 	Test Passed
allocate 10 chunks of memory
Test empty FSPoolCountFreeChunks. 	Test Passed
No memory left in the pool!
Test  FSPoolCountFreeChunks. 		Test Passed
Test  FSPoolCountFreeChunks. 		Test Passed
Test full FSPoolCountFreeChunks. 	Test Passed
